lib/system-flake.nix

```diff
@@ -72,43 +89,192 @@ let
+                    # tell home-manager to use the global (as in NixOS system-level) pkgs and
+                    # install all  user packages through the users.users.<name>.packages.
+                    home-manager.useGlobalPkgs = true;
+                    home-manager.useUserPackages = true;
+
+                    home-manager.extraSpecialArgs = {
+                      inherit soxin soxincfg home-manager;
+
+                      # the mode allows us to tell at what level we are within the modules.
+                      mode = "home-manager";
+                    }
+                    # include the global special arguments.
+                    // globalSpecialArgs
+                    # include the home-manager special arguments.
+                    // hmSpecialArgs;
+
+                    home-manager.sharedModules =
+                      # include the global modules
+                      extraGlobalModules
+                      # include the home-manager modules
+                      ++ extraHomeManagerModules
+                      # include Soxin module
+                      ++ (singleton soxin.nixosModule);
+                  })
+                ;
+              }
+            ))
+          darwinOnlyHosts;
+
+      nixosHosts =
+        let
+          nixosOnlyHosts = filterAttrs (n: host: host.mode == "NixOS") hosts;
+        in
+        mapAttrs
+          (_: host: (recursiveUpdate
+            # pass along the hosts minus few keys that are implementation detail to soxin.
+            (removeAttrs host [ "deploy" "mode" ])
+
+            {
+              specialArgs =
+                {
+                  inherit
+                    home-manager
+                    soxin
+                    soxincfg
+                    ;
+
+                  inherit (host)
+                    mode
+                    ;
+                }
+                # include the specialArgs that were passed in.
+                // (host.specialArgs or { })
+                # include the global special arguments.
+                // globalSpecialArgs
+                # include the NixOS special arguments.
+                // nixosSpecialArgs
+              ;
+
+              modules =
+                (host.modules or [ ])
+                # include sops
+                ++ (optionals withSops (singleton sops-nix.nixosModules.sops))
+                # include sane flake defaults from flake-utils-plus which sets sane `nix.*` defaults.
+                # Please refer to implementation/readme in
+                # github:gytis-ivaskevicius/flake-utils-plus for more details.
+                ++ (singleton flake-utils-plus.nixosModules.saneFlakeDefaults)
+                # include the NixOS modules
+                ++ extraNixosModules
+                # include Soxin modules
+                ++ (singleton soxin.nixosModule)
+                # include home-manager modules
+                ++ (singleton home-manager.nixosModules.home-manager)
```

Comment:
I wonder if there is a way to factorize all of that code.
Somewhere along the lines of:
```
commonModulesBases = {
  sops-nix = "sops";
  flake-utils-plus = "saneFlakeDefaults";
  extraModules = "modules";
  /* ... */
};

commonModulesDarwin = mapAttrsToList (name: value: inputs.${name}.darwinModules.${value});
commonModulesNixOS = mapAttrsToList (name: value: inputs.${name}.nixosModules.${value});
```

what do you think?

Reply:
Yes, I agree but thought of doing a refactor later as it's best to get the functionality working correctly first.